File: 393. UTF-8 Validation.py
class Solution(object):
    def validUtf8(self, data):
        """
        :type data: List[int]
        :rtype: bool
        """
        i = 0
        n = len(data)
        while i < n:
            if data[i] & 0xf8 == 0xf0:
                j = 3
            elif data[i] & 0xf0 == 0xe0:
                j = 2
            elif data[i] & 0xe0 == 0xc0:
                j = 1
            elif data[i] & 0x80 == 0:
                j = 0
            else:
                return False
            print 'i, j = ', i, j
            for k in xrange(j):
                i += 1
                if i >= n:
                    return False
                if data[i] & 0xc0 != 0x80:
                    return False
            i += 1
            print 'i = ', i
                
                
                
        
        return i == n
        
        

File: 393. UTF-8 Validation.go
func validUtf8(data []int) bool {
	n := len(data)
	i := 0
	for i < n {
		// 4 bytes
		if data[i]&0xf8 == 0xf0 {
			if i+4 > n || data[i+1]&0xc0 != 0x80 || data[i+2]&0xc0 != 0x80 || data[i+3]&0xc0 != 0x80 {
				return false
			}
			i += 4
		} else if data[i]&0xf0 == 0xe0 {
			if i+3 > n || data[i+1]&0xc0 != 0x80 || data[i+2]&0xc0 != 0x80 {
				return false
			}
			i += 3
		} else if data[i]&0xe0 == 0xc0 {
			if i+2 > n || data[i+1]&0xc0 != 0x80 {
				return false
			}
			i += 2
		} else if data[i]&0x80 == 0 {
			i++
		} else {
			return false
		}
	}
	return true
}
